Lib/concurrent/futures/process.py

```diff
@@ -69,26 +69,37 @@ class _ThreadWakeup:
     def __init__(self):
         self._closed = False
         self._reader, self._writer = mp.Pipe(duplex=False)
+        self._awaken = False
 
     def close(self):
         # Please note that we do not take the shutdown lock when
         # calling clear() (to avoid deadlocking) so this method can
         # only be called safely from the same thread as all calls to
         # clear() even if you hold the shutdown lock. Otherwise we
         # might try to read from the closed pipe.
-        if not self._closed:
-            self._closed = True
-            self._writer.close()
-            self._reader.close()
+        if self._closed:
+            return
+        self._closed = True
+        self._writer.close()
+        self._reader.close()
 
     def wakeup(self):
-        if not self._closed:
-            self._writer.send_bytes(b"")
+        if self._closed:
+            return
+        if self._awaken:
+            # gh-105829: Write a single byte to not block if the pipe is full.
+            # wait_result_broken_or_wakeup() does not even read the written
+            # byte. It only check if something was written to the pipe.
+            return
+        self._writer.send_bytes(b"x")
```

Comment:
I don't get how it was possible to wake up the pipe before, if no data is written.

Reply:
An empty bytes object is still a thing. It will cause 4 bytes to be sent on the pipe (A zero encoding the length of the byte string followed by zero bytes of data). This will send 5 bytes.

Reply:
Oh ok. I suppose that writing 4 or 5 bytes doesn't make any big difference, as soon as we only send 5 bytes and don't fill the pipe.